// @sanity/runtime-cli/dist/commands/functions/env/list.js
import { Args } from '@oclif/core';
import { DeployedBlueprintCommand } from '../../../baseCommands.js';
import { functionEnvListCore } from '../../../cores/functions/env/list.js';
export default class EnvListCommand extends DeployedBlueprintCommand {
    static args = {
        name: Args.string({ description: 'The name of the Sanity Function', required: true }),
    };
    static description = 'List the environment variables for a Sanity function';
    static examples = ['<%= config.bin %> <%= command.id %> MyFunction'];
    async run() {
        const { success, error } = await functionEnvListCore({
            bin: this.config.bin,
            log: (msg) => this.log(msg),
            args: this.args,
            auth: this.auth,
            blueprint: this.blueprint,
            deployedStack: this.deployedStack,
            scopeType: this.scopeType,
            scopeId: this.scopeId,
            token: this.sanityToken,
            stackId: this.stackId,
        });
        if (!success)
            this.error(error);
    }
}
