// @sanity/runtime-cli/dist/commands/functions/dev.js
import { Flags } from '@oclif/core';
import { BlueprintCommand } from '../../baseCommands.js';
import { functionDevCore } from '../../cores/functions/dev.js';
export default class DevCommand extends BlueprintCommand {
    static description = 'Start the Sanity Function emulator';
    static examples = ['<%= config.bin %> <%= command.id %> --host 127.0.0.1 --port 8974'];
    static flags = {
        host: Flags.string({
            char: 'h',
            description: 'The local network interface at which to listen. [default: "localhost"]',
            required: false,
        }),
        port: Flags.integer({
            char: 'p',
            description: 'TCP port to start emulator on. [default: 8080]',
            required: false,
        }),
        timeout: Flags.integer({
            char: 't',
            description: 'Maximum execution time for all functions, in seconds. Takes precedence over function-specific `timeout`',
            required: false,
        }),
    };
    async run() {
        const { flags } = await this.parse(DevCommand);
        const { success, error, streaming } = await functionDevCore({
            bin: this.config.bin,
            log: (msg) => this.log(msg),
            flags,
        });
        if (!success)
            this.error(error);
        if (streaming)
            await streaming;
    }
}
