// @sanity/runtime-cli/dist/commands/blueprints/config.js
import { Flags } from '@oclif/core';
import { BlueprintCommand } from '../../baseCommands.js';
import { blueprintConfigCore } from '../../cores/blueprints/config.js';
export default class ConfigCommand extends BlueprintCommand {
    static description = 'View or edit Blueprint configuration';
    static examples = [
        '<%= config.bin %> <%= command.id %>',
        '<%= config.bin %> <%= command.id %> --edit',
        '<%= config.bin %> <%= command.id %> --edit --project-id <projectId>',
        '<%= config.bin %> <%= command.id %> --edit --project-id <projectId> --stack-id <stackId>',
    ];
    static flags = {
        edit: Flags.boolean({
            char: 'e',
            description: 'Modify the configuration interactively, or directly when combined with ID flags.',
            default: false,
        }),
        'project-id': Flags.string({
            description: 'Directly set the Project ID in the configuration. Requires --edit flag',
            aliases: ['project', 'projectId'],
            dependsOn: ['edit'],
        }),
        'organization-id': Flags.string({
            description: 'Directly set the Organization ID in the configuration. Requires --edit flag',
            aliases: ['organization', 'organizationId', 'org'],
            dependsOn: ['edit'],
        }),
        'stack-id': Flags.string({
            description: 'Directly set the Stack ID in the configuration. Requires --edit flag',
            aliases: ['stack', 'stackId'],
            dependsOn: ['edit'],
        }),
    };
    async run() {
        const { success, error } = await blueprintConfigCore({
            bin: this.config.bin,
            log: (message) => this.log(message),
            blueprint: this.blueprint,
            token: this.sanityToken,
            flags: this.flags,
        });
        if (!success)
            this.error(error);
    }
}
