// @sanity/runtime-cli/dist/commands/blueprints/deploy.js
import { Flags } from '@oclif/core';
import { DeployedBlueprintCommand } from '../../baseCommands.js';
import { blueprintDeployCore } from '../../cores/blueprints/deploy.js';
export default class DeployCommand extends DeployedBlueprintCommand {
    static description = 'Deploy a Blueprint';
    static examples = [
        '<%= config.bin %> <%= command.id %>',
        '<%= config.bin %> <%= command.id %> --no-wait',
    ];
    static flags = {
        'no-wait': Flags.boolean({
            description: 'Do not wait for Stack deployment to complete',
            default: false,
        }),
    };
    async run() {
        const { success, error } = await blueprintDeployCore({
            bin: this.config.bin,
            log: (message) => this.log(message),
            auth: this.auth,
            stackId: this.stackId,
            scopeType: this.scopeType,
            scopeId: this.scopeId,
            deployedStack: this.deployedStack,
            blueprint: this.blueprint,
            flags: this.flags,
        });
        if (!success)
            this.error(error);
    }
}
