// @sanity/runtime-cli/dist/commands/blueprints/stacks.js
import { Flags } from '@oclif/core';
import { BlueprintCommand } from '../../baseCommands.js';
import { blueprintStacksCore } from '../../cores/blueprints/stacks.js';
export default class StacksCommand extends BlueprintCommand {
    static description = 'List all Blueprint Stacks';
    static examples = [
        '<%= config.bin %> <%= command.id %>',
        '<%= config.bin %> <%= command.id %> --project-id <projectId>',
        '<%= config.bin %> <%= command.id %> --organization-id <organizationId>',
    ];
    static flags = {
        'project-id': Flags.string({
            description: 'Project ID to show stacks for',
            aliases: ['projectId', 'project'],
            exclusive: ['organization-id'],
        }),
        'organization-id': Flags.string({
            description: 'Organization ID to show stacks for',
            aliases: ['organizationId', 'organization', 'org'],
            exclusive: ['project-id'],
        }),
    };
    async run() {
        const { success, error } = await blueprintStacksCore({
            bin: this.config.bin,
            log: (message) => this.log(message),
            token: this.sanityToken,
            blueprint: this.blueprint,
            flags: this.flags,
        });
        if (!success)
            this.error(error);
    }
}
