// @sanity/cli-core/dist/config/studio/readStudioConfig.js
import { dirname } from 'node:path';
import { z } from 'zod';
import { studioWorkerTask } from '../../loaders/studio/studioWorkerTask.js';
const schemaSchema = z.object({
    name: z.string().optional(),
    types: z.array(z.object({}).passthrough())
});
const singleStudioWorkspaceSchema = z.object({
    basePath: z.string().optional(),
    dataset: z.string(),
    name: z.string().optional(),
    plugins: z.array(z.unknown()).optional(),
    projectId: z.string(),
    schema: schemaSchema.optional(),
    title: z.string().optional()
}).passthrough();
const studioWorkspaceSchema = z.object({
    basePath: z.string(),
    dataset: z.string(),
    name: z.string(),
    plugins: z.array(z.unknown()).optional(),
    projectId: z.string(),
    schema: z.object({
        _original: schemaSchema
    }),
    title: z.string()
});
const rawConfigSchema = z.union([
    z.array(studioWorkspaceSchema),
    singleStudioWorkspaceSchema
]);
const resolvedConfigSchema = z.array(studioWorkspaceSchema);
export async function readStudioConfig(configPath, options) {
    const result = await studioWorkerTask(new URL('readStudioConfig.worker.js', import.meta.url), {
        name: 'studioConfig',
        studioRootPath: dirname(configPath),
        workerData: {
            configPath,
            resolvePlugins: options.resolvePlugins
        }
    });
    return options.resolvePlugins ? resolvedConfigSchema.parse(result) : rawConfigSchema.parse(result);
}
