// @sanity/runtime-cli/dist/server/static/api.js
/* eslint-disable n/no-unsupported-features/node-builtins */
import {Store} from './vendor/vendor.bundle.js'

// list of events to simulate
const events = ['create', 'update', 'delete'].map((e) => ({name: e}))

// eslint-disable-next-line new-cap
const store = Store({events, selectedEvent: events[0].name})

export default function API() {
  return {
    blueprint,
    document,
    invoke,
    projects,
    datasets,
    organizations,
    mediaLibraries,
    asset,
    store,
    subscribe: store.subscribe,
    unsubscribe: store.unsubscribe,
  }
}

function invoke({context, event, metadata}) {
  store.inprogress = true
  const start = Date.now()
  const payload = {
    data: {context, event},
    func: store.selectedIndex,
    metadata,
  }
  fetch('/invoke', {
    body: JSON.stringify(payload),
    headers: {
      'Content-Type': 'application/json',
    },
    method: 'POST',
  })
    .then((response) => {
      return response.json().then((data) => ({data, timings: getServerTimings(response)}))
    })
    .then(({data, timings}) => {
      store.inprogress = false
      store.result = {
        ...data,
        time: Date.now() - start,
        timings,
      }
    })
}

function blueprint() {
  fetch('/blueprint')
    .then((response) => response.json())
    .then((blueprint) => {
      const {parsedBlueprint, projectId, organizationId} = blueprint
      const functions = parsedBlueprint?.resources.filter((r) =>
        r.type.startsWith('sanity.function.'),
      )

      store.functions = functions
      store.selectedIndex = functions[0].name
      store.selectedFunctionType = functions[0].type
      store.defaultProject = projectId || undefined
      store.defaultOrganization = organizationId || undefined
    })
    .catch(() => {
      store.functions = []
    })
}

function projects() {
  fetch('/projects')
    .then((response) => response.json())
    .then(async (projects) => {
      store.projects = projects
      store.selectedProject = store.defaultProject ? store.defaultProject : projects[0].id
      await datasets(store.selectedProject)
    })
    .catch(() => {
      store.projects = []
    })
}

function datasets(selectedProject) {
  fetch(`/datasets?project=${selectedProject}`)
    .then((response) => response.json())
    .then((datasets) => {
      store.datasets = datasets
      store.selectedDataset = datasets[0].name
    })
    .catch(() => {
      store.datasets = []
    })
}

function document({projectId, dataset, docId}) {
  return fetch(`/document?project=${projectId}&dataset=${dataset}&doc=${docId}`)
    .then((response) => response.json())
    .then((doc) => {
      store.document = doc
    })
    .catch(() => {
      store.document = {}
    })
}

function organizations() {
  fetch('/organizations')
    .then((response) => response.json())
    .then(async (organizations) => {
      store.organizations = organizations
      store.selectedOrganization = store.defaultOrganization
        ? store.defaultOrganization
        : organizations[0].id
      await mediaLibraries(store.selectedOrganization)
    })
    .catch(() => {
      store.organizations = []
    })
}

function mediaLibraries(selectedOrganization) {
  fetch(`/media-libraries?organization=${selectedOrganization}`)
    .then((response) => response.json())
    .then((mediaLibraries) => {
      store.mediaLibraries = mediaLibraries
      store.selectedMediaLibrary = store.defaultMediaLibrary
        ? store.defaultMediaLibrary
        : mediaLibraries[0].id
    })
    .catch(() => {
      store.mediaLibraries = []
    })
}

function asset({organizationId, mediaLibraryId, docId}) {
  return fetch(`/asset?organization=${organizationId}&medialibrary=${mediaLibraryId}&doc=${docId}`)
    .then((response) => response.json())
    .then((doc) => {
      store.document = doc
    })
    .catch(() => {
      store.document = {}
    })
}

function getServerTimings(response) {
  const timings = {}
  const serverTiming = response.headers.get('Server-Timing')
  if (!serverTiming) {
    return timings
  }

  for (const entry of serverTiming.split(',')) {
    const [name, ...params] = entry.split(';')
    const durationParam = params.find((p) => p.startsWith('dur='))
    if (durationParam) {
      timings[name.trim()] = Number.parseFloat(durationParam.slice(4))
    }
  }

  return timings
}
