// @sanity/runtime-cli/dist/utils/display/presenters.js
import { cwd } from 'node:process';
import chalk from 'chalk';
export function check(str) {
    return `${chalk.bold(chalk.green('✔︎'))} ${str}`;
}
export function info(str) {
    return `${chalk.bold.blue('ℹ︎')} ${str}`;
}
export function warn(str) {
    return `${chalk.bold.yellow('▶︎')} ${str}`;
}
export function unsure(str) {
    return `${chalk.bold.cyan('?')} ${str}`;
}
export function severe(str) {
    return `${chalk.bold.red('✘')} ${str}`;
}
export function niceId(id) {
    if (!id)
        return '';
    return `<${chalk.yellow(id)}>`;
}
export function indent(str, spaces = 2) {
    const pad = ' '.repeat(spaces);
    return str
        .split('\n')
        .map((line) => (line.length > 0 ? pad + line : line))
        .join('\n');
}
export function capitalize(str) {
    return str.charAt(0).toUpperCase() + str.slice(1);
}
export function filePathRelativeToCwd(filePath) {
    return filePath.replace(cwd(), '.');
}
export function labeledId(label, id) {
    return `${chalk.blue(capitalize(label || 'unknown'))} ${niceId(id || 'unknown')}`;
}
